main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala

```diff
@@ -0,0 +1,243 @@
+      * @param pair The entry which to make uncurried versions of
+      * @return A new map with the uncurried versions added
+      */
+    def mkUncurriedDefs(map: Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant], pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      map ++ mkUncurriedDef(pair)
+    }
+
+    /**
+      * Creates all the uncurried versions of a function
+      *
+      * @param pair The dfnsym definition pair to base the functions on
+      * @return A map containing all the new definitions
+      */
+    def mkUncurriedDef(pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      val sym = pair._1
+      val curriedDef = pair._2
+      curriedDef.exp match {
+        // If the body is a lambda, then we have a curried function
+        // definition, so we need to create an uncurried version.
+        // To do this, we'll take make the body of the function the body
+        // of the lambda, then add the arguments of the lambda to the
+        // formal parameters
+        case Lambda(args, body, _, _) =>
+          // Create a new definition for our function
+          val uncurriedSym = Symbol.mkDefnSym(pair._1.name + "$uncurried")
+          // Create new VarSyms for our function
+          val newFormals = (curriedDef.formals ::: args).map(f => f.copy(sym = Symbol.freshVarSym(f.sym)))
+          // Replace the old VarSyms in the body
+          val varMapping = (curriedDef.formals ::: args).map(f => f.sym).zip(newFormals.map(f => f.sym)).toMap
+          val newBody = Expressions.replaceVars(body, varMapping)
+
+          val unCurriedDef = SimplifiedAst.Definition.Constant(
+            curriedDef.ann,
+            curriedDef.mod,
+            uncurriedSym,
+            newFormals,
+            newBody,
+            curriedDef.isSynthetic,
+            uncurryType(curriedDef.tpe),
+            curriedDef.loc)
+          (Map.empty[Symbol.DefnSym, SimplifiedAst.Definition.Constant] + pair + (uncurriedSym -> unCurriedDef)) ++ mkUncurriedDef(uncurriedSym, unCurriedDef)
+        case _ => Map.empty[Symbol.DefnSym, SimplifiedAst.Definition.Constant] + pair
+      }
+    }
+
+    /**
+      * Uncurry a definition constant
+      *
+      * @param cst
+      * @param root
+      * @return
+      */
+    def uncurry(cst: SimplifiedAst.Definition.Constant, root: SimplifiedAst.Root): SimplifiedAst.Definition.Constant = {
+      cst.copy(exp = Expressions.uncurry(cst.exp, root))
+    }
+  }
+
+  object Expressions {
+    /**
+      * Recursively replace all the variables in tast with new variables using the
+      * pairs in replaceSyms. This ensure that the new functions created don't
+      * reference the variables of the original functions.
+      *
+      * @param oldExp The expression to replace
```

Comment:
Skip the documentation for parameters. Just focus on the documentation for the overall function.